sync_test.go

```diff
@@ -1034,78 +1049,272 @@ func testRandomBlocks(harness *neutrinoHarness, t *testing.T) {
+// TestNeutrinoSyncWithHeadersImport tests the functionality of Neutrino's
+// headers import feature. The test creates a btcd SimNet node, generates
+// blocks, and syncs a temporary Neutrino instance with this node to obtain
+// headers. It then prepares header files for import by copying them and adding
+// the necessary metadata.
+//
+// The first part of the test creates a new Neutrino instance that imports
+// these headers from files without connecting to any peers. It verifies
+// that headers are correctly imported and the chain is properly synchronized
+// without network assistance.
+//
+// The second part generates additional blocks and creates another Neutrino
+// instance with the same import configuration but with network connectivity.
+// This tests that Neutrino correctly handles the case where header stores are
+// already populated, skipping the import process and continuing synchronization
+// from its current state.
+//
+// This test validates that the headers import mechanism works correctly both as
+// a way to bootstrap new nodes and when restarting existing nodes,
+// significantly improving sync performance in the former scenario.
+func TestNeutrinoSyncWithHeadersImport(t *testing.T) {
+	// Setup context during testing.
+	rootCtx := context.Background()
 
 	// Create a btcd SimNet node and generate 800 blocks
 	h1, err := rpctest.New(
 		&chaincfg.SimNetParams, nil, []string{"--txindex"}, "",
 	)
-	if err != nil {
-		t.Fatalf("Couldn't create harness: %s", err)
-	}
+	require.NoError(t, err)
 	defer h1.TearDown()
+
 	err = h1.SetUp(false, 0)
-	if err != nil {
-		t.Fatalf("Couldn't set up harness: %s", err)
-	}
+	require.NoError(t, err)
+
 	_, err = h1.Client.Generate(800)
-	if err != nil {
-		t.Fatalf("Couldn't generate blocks: %s", err)
+	require.NoError(t, err)
+
+	// Create a temporary directory for the headers data.
+	tempDir := t.TempDir()
+	defer os.RemoveAll(tempDir)
+
+	// Create filters db for indexing the headers.
+	db, err := walletdb.Create(
+		"bdb", tempDir+"/filters.db", true, dbOpenTimeout,
+	)
+	require.NoError(t, err)
+	defer db.Close()
+
+	config := neutrino.Config{
+		DataDir:     tempDir,
+		Database:    db,
+		ChainParams: chaincfg.SimNetParams,
+		AddPeers: []string{
+			h1.P2PAddress(),
+		},
+		HeadersImport: nil,
 	}
 
+	neutrino.MaxPeers = 1
+	neutrino.BanDuration = 5 * time.Second
+	neutrino.QueryPeerConnectTimeout = 10 * time.Second
+
+	// Set up export service to connect to h1 and download headers.
+	exportSvc, err := neutrino.NewChainService(config)
+	require.NoError(t, err)
+	exportSvc.Start(rootCtx)
+
+	testHarness := &neutrinoHarness{
+		h1:  h1,
+		h2:  nil,
+		h3:  nil,
+		svc: exportSvc,
+	}
+
+	// Wait for export service to sync with h1.
+	t.Logf("Waiting for export service to sync with h1...")
+	testInitialSync(testHarness, t)
+
+	// Create directory for importing headers.
+	importDir := t.TempDir()
+	defer os.RemoveAll(importDir)
+
+	blockHeadersFileName := "block_headers.bin"
```

Comment:
I think we could save the files from a run of this test case and with that do some additional unit and benchmark tests directly in the `chainimport` package.

Reply:
No, I was thinking of general benchmark tests of the new code. To see where things could potentially be optimized (e.g. memory allocations, CPU cycles). Not comparing against the old method.